File: 6.z-字形变换.java
/*
 * @lc app=leetcode.cn id=6 lang=java
 *
 * [6] Z 字形变换
 */

// @lc code=start
class Solution {
    static int N = 1010;
    static char[][] g = new char[N][N];
    static int[] idxs = new int[N];
    public String convert(String s, int m) {
        if (m == 1) return s;
        int n = s.length();
        Arrays.fill(idxs, 0);
        for (int i = 0, j = 0, k = 1; i < n; i++) {
            g[j][idxs[j]++] = s.charAt(i);
            j += k;
            if (j < 0) {
                j += 2; k = 1;
            } else if (j == m) {
                j -= 2; k = -1;
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < idxs[i]; j++) {
                sb.append(g[i][j]);
            }
        }
        return sb.toString();
    }
}
// @lc code=end

File: 6.z-字形变换.cpp
/*
 * @lc app=leetcode.cn id=6 lang=cpp
 *
 * [6] Z 字形变换
 */

// @lc code=start
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1) return s;
        vector<string> rows(min(int(s.length()), numRows)); //记录字符串的数组
        int num = 0; //行号
        bool goDown = false; //判断是否往下一行移动
        for (char c : s) {//开始迭代s里的字符
            rows[num] += c; //把字符存入相应的行号
            if (num == 0 || num == numRows - 1) goDown =! goDown; //如果到了行头或者行尾就开始转行
            num += goDown ? 1 : -1; 
        }

        //开始从上往下遍历
        string r;
        for (string row : rows) r += row;
        return r;
    }
};
// @lc code=end

